### PR TITLE
fix: rmdir没有找到目录时不打印日志

### PR DESCRIPTION
**Pull request by 1037827920**
## **Type**
Bug fix, Enhancement

___

## **Description**
此拉取请求修复了`rmdir`命令在找不到指定目录时不打印日志的问题，并增强了该命令的错误处理。现在，如果尝试删除的目录不存在，将返回一个清晰的`DirectoryNotFound`错误。

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix
enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mod.rs
</strong><dd><code>修复并增强`rmdir`命令的错误处理</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/shell/command/mod.rs
- 修复了`rmdir`命令在找不到目录时不打印日志的问题。
- 增强了`rmdir`命令的错误处理，现在如果目录不存在，将返回`DirectoryNotFound`错误。
    </details>
  </td>
  <td><a href="https://github.com/DragonOS-Community/NovaShell/pull/29/files#diff-6466ef2ec51e3d2a0d4452bf3ebbb45533195c015b37e5fcbf1f7c9b446b55dd"></a></td>
</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: src/shell/command/mod.rs [hunk 1]
```diff
@@ -417,7 +417,36 @@ impl Shell {
         if unlikely(args.len() != 1) {
             return Err(CommandError::WrongArgumentCount(args.len()));
         }
-        let path = self.is_dir(args.get(0).unwrap())?;
+
+        let path = args.get(0).unwrap();
```
Comment (fslongjin) [inline src/shell/command/mod.rs]: 不能直接unwrap，否则将会造成shell的panic。应当返回错误。

File: src/shell/command/mod.rs [hunk 2]
```diff
@@ -417,7 +417,36 @@ impl Shell {
...
+        let mut parent = Self::current_dir();
+        let mut child = path.clone();
+        if let Some(index) = path.rfind('/') {
+            let (str1, str2) = path.split_at(index + 1);
+            parent = String::from(str1);
+            child = String::from(str2);
+        }
+
+        let dir = match fs::read_dir(Path::new(&parent)) {
+            Ok(readdir) => readdir,
```
Comment (fslongjin) [inline src/shell/command/mod.rs]: 使用ok_or方法，而不是这种match

File: src/shell/command/mod.rs [hunk 3]
```diff
@@ -417,7 +417,36 @@ impl Shell {
...
+            Err(_) => return Err(CommandError::InvalidArgument(parent)),
+        };
+
+        let mut is_find = false;
+        for entry in dir {
+            let entry = entry.unwrap();
+            if entry.file_type().unwrap().is_dir()
```
Comment (fslongjin) [inline src/shell/command/mod.rs]: 同样的问题，不能直接unwrap。要对空的情况做处理。

**Review by fslongjin — CHANGES_REQUESTED** [review]

File: src/shell/command/mod.rs (hunk 2 above)
Comment (1037827920) [reply src/shell/command/mod.rs]: 不使用match的话，ok_or要取出Ok里面的值只能使用unwrap，但是unwrap返回的Result类型跟函数返回的不一样，还是用match比较好吧

**Comment by fslongjin** [comment]
你可以让gpt帮你改写一下，因为这种写法以后（估计就这一两个月会加上）肯定会在clippy的检查之中报错的，这个不符合rust的那种最佳实践



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年3月13日(周三) 上午8:21
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [DragonOS-Community/NovaShell] fix: rmdir没有找到目录时不打印日志 (PR #29)




 
@1037827920 commented on this pull request.
 
 
In src/shell/command/mod.rs:
 &gt; @@ -417,7 +417,36 @@ impl Shell {          if unlikely(args.len() != 1) {              return Err(CommandError::WrongArgumentCount(args.len()));          } -        let path = self.is_dir(args.get(0).unwrap())?; + +        let path = args.get(0).unwrap(); +        let mut parent = Self::current_dir(); +        let mut child = path.clone(); +        if let Some(index) = path.rfind('/') { +            let (str1, str2) = path.split_at(index + 1); +            parent = String::from(str1); +            child = String::from(str2); +        } + +        let dir = match fs::read_dir(Path::new(&amp;parent)) { +            Ok(readdir) =&gt; readdir,  
不使用match的话，ok_or要取出Ok里面的值只能使用unwrap，但是unwrap返回的Result类型跟函数返回的不一样，还是用match比较好吧

File: src/shell/command/mod.rs (hunk 2 above)
Comment (fslongjin) [reply src/shell/command/mod.rs]: > unwrap返回的Result类型跟函数返回的不一样

Result类型err部分不一样的话就用map_err。 而且ok_or之后再加`?`就能自动把错误值给抛出去

Comment (1037827920) [reply src/shell/command/mod.rs]: ok

**Pull request closed without merging**